src/storage/secondary/RedisStorage.hpp

```diff
@@ -0,0 +1,64 @@
+// Copyright (C) 2021 Joel Rosdahl and other contributors
+//
+// See doc/AUTHORS.adoc for a complete list of contributors.
+//
+// This program is free software; you can redistribute it and/or modify it
+// under the terms of the GNU General Public License as published by the Free
+// Software Foundation; either version 3 of the License, or (at your option)
+// any later version.
+//
+// This program is distributed in the hope that it will be useful, but WITHOUT
+// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+// FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
+// more details.
+//
+// You should have received a copy of the GNU General Public License along with
+// this program; if not, write to the Free Software Foundation, Inc., 51
+// Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
+
+#pragma once
+
+#include <storage/SecondaryStorage.hpp>
+#include <storage/types.hpp>
+
+#pragma GCC diagnostic push
+// ISO C++ forbids flexible array member ‘buf’
+#pragma GCC diagnostic ignored "-Wpedantic"
+// cast increases required alignment from 1 to 4
+#pragma GCC diagnostic ignored "-Wcast-align"
+#include <hiredis.h>
+#pragma GCC diagnostic pop
+
+namespace storage {
+namespace secondary {
+
+class RedisStorage : public storage::SecondaryStorage
+{
+public:
+  RedisStorage(const std::string& url, const AttributeMap& attributes);
+  ~RedisStorage();
+
+  nonstd::expected<nonstd::optional<std::string>, Error>
+  get(const Digest& key) override;
+  nonstd::expected<bool, Error> put(const Digest& key,
+                                    const std::string& value,
+                                    bool only_if_missing) override;
+  nonstd::expected<bool, Error> remove(const Digest& key) override;
+
+private:
+  std::string m_url;
+  redisContext* m_context;
```

Comment:
I'd forward declare the `redisContext` and avoid leaking stuff from `<hiredis.h>` into other consumers of this header. In my http-branch the httplib pulls in `windows.h` things fall apart everywhere.

Reply:
Not sure I would want to expose it being a struct even, but I don't see other ccache headers trying to hide implementation.

Only looked at zstd, though.

Reply:
Sounds good to me to forward declare.

I've written a related to do item for zstd.h as well.